common4j/src/main/com/microsoft/identity/common/java/cache/SharedPreferencesAccountCredentialCache.java

```diff
@@ -100,112 +107,102 @@ public SharedPreferencesAccountCredentialCache(
         Logger.verbose(TAG, "Init: " + TAG);
         mSharedPreferencesFileManager = sharedPreferencesFileManager;
         mCacheValueDelegate = accountCacheValueDelegate;
+        new Thread(() -> load()).start();
+    }
+
+    private void load()
+    {
+        synchronized (mCacheLock) {
+            try {
+                mCachedAccountRecordsWithKeys = loadAccountsWithKeys();
+                Logger.info(TAG, "Loaded " + mCachedAccountRecordsWithKeys.size() + " AccountRecords");
+                mCachedCredentialsWithKeys = loadCredentialsWithKeys();
+                Logger.info(TAG, "Loaded " + mCachedCredentialsWithKeys.size() + " Credentials");
+            } catch (Throwable t) {
+                Logger.error(TAG, "Failed to load initial credentials from SharedPreferences", t);
+            } finally {
+                mLoaded = true;
+                mCacheLock.notifyAll();
+            }
+        }
+    }
+
+    private void waitForInitialLoad() {
+        if (mLoaded) {
+            return;
+        }
+
+        while (!mLoaded) {
+            try {
+                mCacheLock.wait();
+            } catch (InterruptedException e) {}
+        }
     }
 
     @Override
-    public synchronized void saveAccount(@NonNull final AccountRecord accountToSave) {
+    public void saveAccount(@NonNull final AccountRecord accountToSave) {
         Logger.verbose(TAG, "Saving Account...");
         Logger.verbose(TAG, "Account type: [" + accountToSave.getClass().getSimpleName() + "]");
         final String cacheKey = mCacheValueDelegate.generateCacheKey(accountToSave);
         Logger.verbosePII(TAG, "Generated cache key: [" + cacheKey + "]");
 
-        // Perform any necessary field merging on the Account to save...
-        final AccountRecord existingAccount = getAccount(cacheKey);
+        synchronized (mCacheLock) {
```

Comment:
should we check for mLoaded outside of the `synchronized` block to avoid it,
I think that way we can avoid the overhead of making it `synchronized` 

Same in other places.